### ISSUE TITLE
BUG: Using solution from Differential Evolution directly in optimization function yields drastically different results

### ISSUE THREAD

**Issue by Chandlertso** [scipy/scipy#15594]
### Describe your issue.

Hi!  Sorry for any awkward wording, I'm a novice and entirely self-taught.

I have a function which I'm using differential evolution to find the optimal values for, using 17 variables.  I can't post the code because it's proprietary and I can't really recreate it because it would take a while because it's kinda weird and I'm somewhat new to this.  

WhenI run scipy's DE method on my function, I get the solution and paramaters which yielded that exact solution.  However, whenever I plug those parameters directly into the same exact function I was optimizing for, I get a very different result.  

There are a number of things that I think can be causing this issue.
1. I am using workers: -1, and with 32 useable cores, maybe this causes some issue.
2.  in my objective function, I call another function which is vectorized and contains about 16 different if statements and some calcuations. 
3. I am using the wrapdisc library to set bounds and limit variables to a certain decimal length
4. any combination of those
5. I'm just bad and there really isn't a bug

I'm using 1.8 and everything else is up to date.  Python 3.10

I've tried this now with two different machines two, both running windows.
I will try running it with 1 worker right now (this will take a day to finish).

There are no error messages and the rest of the code runs correctly to the best of my knowledge.

I can share my code directly with an individual if they can help, just cannot post directly.

edit:
I've tried using best1bin and randtobest1bin.  I'll be testing all of the options out shortly. 

edit2:
I've been storing the results from nearly a hundred different optimizations that have been run.  All of them were from scipy 1.73.  All of them are showing the same problems.  
I just finished running with randtobest1bin and that had the same problem, but the numbers were much closer than the others.  still testing all the different strategies.

edit3:
I'm currently working on two projects which use scipy's DE.  I just checked and the other project does not have this problem.

### Reproducing Code Example

```python
N/A
```

### Error message

```shell
N/A
```

### SciPy/NumPy/Python version information

1.8

**Comment by andyfaff** [comment]
It sounds like you're saying that the following assertion would fail: 
```
np.testing.assert_equal(fun(OptimizeResult.x), OptimizeResult.fun) 
```

This circumstance is already [being examined](https://github.com/scipy/scipy/blob/266670d2674f39c06b934c0ba6ee4d0f76ebd9e9/scipy/optimize/tests/test__differential_evolution.py#L231) in the unit tests.

**Comment by tupui** [comment]
> WhenI run scipy's DE method on my function, I get the solution and paramaters which yielded that exact solution. However, whenever I plug those parameters directly into the same exact function I was optimizing for, I get a very different result.

Can you elaborate on your function? DE is stochastic meaning that it will not necessarily follow the same path during the optimization. You can try to use `seed` and check that you get the same result. If not, it could indicate an issue on your end. Or your function is not deterministic (same parameters lead to same output).

**Comment by Chandlertso** [comment]
Thank you for the quick response!
So I tried the assert.equal function and it failed.  here is the output:

  File "C:\Users\chand\AppData\Local\Programs\Python\Python310\lib\site-packages\numpy\testing\_private\utils.py", line 425, in assert_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
 ACTUAL: -1709935435.93
 DESIRED: -2710952117.96

I'll try using seed.  All I have to do is set seed equal to an int and run it multiple times right?  like: seed = 1,  
and then I compare the results?

Also, I'm pretty sure it's deterministic, the same parameters lead to the same output.  The function looks at a bunch of numbers in a list and does some calculations.  There is no random element in there. 

I just got done trying out all of the different strategies and init options.  best1bin was the worst with a 35% difference between fun(OptimizeResult.x), OptimizeResult.fun  .  best2bin only has a .1% difference.  randtobest1exp had a 5% difference.  I can share the rest of my results if you're interested.  I ran all of them with init = sobol. 

I ran it three times with best1bin and initi = sobol, hypercube, and halton.  hypercuber was 10% off, halton was 20% off, and sobol was 35% off.  I wonder if pop size has an impact in all of this. 

Currently running it with workers = 1. will take at least a day for it to finish.

**Comment by tupui** [comment]
> Thank you for the quick response! So I tried the assert.equal function and it failed. here is the output:
> 
> File "C:\Users\chand\AppData\Local\Programs\Python\Python310\lib\site-packages\numpy\testing_private\utils.py", line 425, in assert_equal raise AssertionError(msg) AssertionError: Items are not equal: ACTUAL: -1709935435.93 DESIRED: -2710952117.96
> 
> I'll try using seed. All I have to do is set seed equal to an int and run it multiple times right? like: seed = 1, and then I compare the results?

Yes.

> I just got done trying out all of the different strategies and init options. best1bin was the worst with a 35% difference between fun(OptimizeResult.x), OptimizeResult.fun . best2bin only has a .1% difference. randtobest1exp had a 5% difference. I can share the rest of my results if you're interested. I ran all of them with init = sobol.

All of that is expected, things like Sobol', Halton (even Latin Hypercube) produces non deterministic sequences of numbers. So everytime you restart, you will have a different population, hence a different optimization path.

> I ran it three times with best1bin and initi = sobol, hypercube, and halton. hypercuber was 10% off, halton was 20% off, and sobol was 35% off. I wonder if pop size has an impact in all of this.

Yes the population size has an impact since the parameter space is more or less explored at each iterations.

The method is stochastic by nature, a 0.1% is really not much and even in the other cases is really problem dependent and if you have a noisy function can be normal. Based on that I think it's safe to say that there is no issue. I will wait for your test with the seed, but I think we can close this.

**Comment by Chandlertso** [comment]
Okay, so I don't think I'm understanding something here.  What I've been doing is comparing the numbers from the DE OptimizeResult.fun against the fun(OptimizeResult.x) for all the different strategies, and what I found is that some strategies have produce a OptimizeResult.fun number that is significantly closer to the fun(OptimizeResult.x) than others.  

Put very simply (because I don't have faith in my ability to communicate these things using proper terminology) I run the differential_evolution(myfunction, bounds = bounds).  take X, the array that shows the optimal parameters, and then plug it back into myfunction().  I then compared the results from the two.  

I ever so slightly understand the difference between the strategies and to me, it shouldn't impact the differences between the myfunction(x) and optimizedresult.fun  because the strategy is just changing the method in which the DE algo chooses the next generation / offspring, it's not changing how the objective function is being graded or determined.

**Comment by rkern** [comment]
I suspect something is going on with how you are using `wrapdisc` when you go to make the final comparison. I see from its [README](https://github.com/impredicative/wrapdisc/) that you have to decode the `result.x` before passing it into your unwrapped objective function. It's not clear to me which you are comparing to each other. Maybe there's an issue with its cache, too. [See this note](https://github.com/impredicative/wrapdisc/blob/master/wrapdisc/wrapdisc.py#L61).

I do recommend going through the process to reduce your example demonstrating the problem down to a complete minimal example that you could post here for us to see and run (it can use `wrapdisc` since I suspect that's the root of your problem). Even if you never do post it here, going through the process of simplification helps you identify the issue.

**Comment by Chandlertso** [comment]
Okay, running the optimization with 1 worker also had the same issue.  It was off by nearly the same amount as worker = -1, controlling for the same strategy and init. 

I'm also controlling for warpdisc's rounding when comparing the OptimizeResult.fun against the fun(OptimizeResult.x).  I've actually been comparing both fun(OptimizeResult.x) and the rounded version of fun(OptimizeResult.x).  

I'm also using wrapdisc in my other project and haven't had a problem with it there.

My guess currently now is with using vectorization within the objective function.  I'll be writing a simplified version of my code soon.

**Comment by andyfaff** [comment]
@Chandlertso, I'm pretty sure this isn't a bug, so I'm going to close this issue. If you can come up with a minimal working example that demonstrates a problem please feel free to re-open the issue.

**Comment by Chandlertso** [comment]
Okay I figured it out!  Something wacky is going on with the wrapdisc library.  Don't know what it is and I'm not going to find out lololol.  So yeah, just so you all know, sometimes wrapdisc makes the DE not work properly.  Ran everything without wrapdisc and it all runs correctly and OptimizeResult.fun = fun(OptimizeResult.x)
Sorry for all the hassle, hopefully if someone else has my problem in the future, you know the solution : )
